File: 63.py
class Solution:
    # @param {integer[][]} obstacleGrid
    # @return {integer}
    def uniquePathsWithObstacles(self, obstacleGrid):
        
        m = len(obstacleGrid)
        n = len(obstacleGrid[0])
        
        if obstacleGrid[0][0] == 1:
            obstacleGrid[0][0] = 0
        else:
            obstacleGrid[0][0] = 1
        
        for i in range(1,m):
            if obstacleGrid[i][0] == 1:
                obstacleGrid[i][0] = 0
            else:
                obstacleGrid[i][0] = obstacleGrid[i-1][0]
        for j in range(1,n):
            if obstacleGrid[0][j] == 1:
                obstacleGrid[0][j] = 0
            else:
                obstacleGrid[0][j] = obstacleGrid[0][j-1]
        
        
        
        for i in range(1,m):
            for j in range(1,n):
                if obstacleGrid[i][j] == 1:
                    obstacleGrid[i][j] = 0
                else:
                    obstacleGrid[i][j] = obstacleGrid[i-1][j] + obstacleGrid[i][j-1]
                    
        return obstacleGrid[m-1][n-1]
        

File: 63.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m = obstacleGrid.size();
        int n = obstacleGrid[0].size();
        vector<int> mat(n,0);
        mat[0] = 1;
        
        for(int i = 0 ; i < m ; i++)
            for(int j = 0; j<n; j++){
                if(obstacleGrid[i][j]==1)
                    mat[j] = 0;
                else if (j > 0)
                    mat[j] = mat[j] + mat[j-1];
           
            }
        
        
        return mat[n-1];
        
    }
};